### ISSUE TITLE
https_web: allow specifying a private CA for the bundle endpoint

### ISSUE THREAD

**Issue by sorindumitru** [spiffe/spiffe#297]
I'd like to be able to use my own CA for the https_web profile of the bundle endpoint.  This is currently not allowed by the specification through [wording such as](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Federation.md#521-web-pki-https_web):
* endpoints using the https_web profile type are authenticated using the same public CA certificates commonly installed in modern operating systems.
* The https_web profile MUST NOT require any additional parameters beyond those required for every profile (namely the trust domain name, profile type, and endpoint URL) in order to function.

Some users prefer to use private CAs instead of a public CA because they have full control over what that certificate is issued. In the WebPKI world there have been additional checks added to certificate validation (e.g. certificate transparency logs) to address those issues, but those might not be done by all clients.

To note also that the requirements above can be achieved with a private CA by placing the CA in the system trust store, but that's not something that can always be done (e.g. easy to do in Docker, harder to do on a shared system).

**Comment by spikecurtis** [comment]
Thanks for opening an issue.

I was wondering if you could describe the motivating use case in some more detail?

In designing the initial profiles, IIRC, we figured that https_web would be for use cases where you didn't have a way to pre-provision the bundle endpoint client with a CA. If you're going to the trouble of minting your own CA, is there some reason you don't want to make it a SPIFFE compatible CA and use the https_spiffe profile?

**Comment by sorindumitru** [comment]
I think it can be much easier to trust a single CA that you own, as opposed to trusting the whole set of CAs trusted for WebPKI. There have been issues in the past with certain authorities. The web pki/browser consortium world has some way of dealing with that, but I don't think all http clients do the extra checks. Also, in some cases it might be hard to get certificates issued by a trusted authority for all of your endpoints, but it may be very easier to get them issued by your own authority.

For the SPIFFE compatible suggestion, are you proposing to make a trust domain with my CA in it and then use it for validing all the endpoints that use certs issued by it? I think that would work, but seems overly complicated for achieving the same thing.

What are the general concerns here? Is it that widening what's allowed on this profile might make it harder to federate trust domains?

**Comment by spikecurtis** [comment]
> For the SPIFFE compatible suggestion, are you proposing to make a trust domain with my CA in it and then use it for validing all the endpoints that use certs issued by it? I think that would work, but seems overly complicated for achieving the same thing.

We're talking about SPIFFE Federation here, so the assumption is that you already have some SPIFFE trust domain with a CA root of trust.  That CA signs an SVID that is issued to the bundle endpoint. Then to authenticate the endpoint, you need the URL, SPIFFE-ID and CA --- very similar to what you're asking for except the certs are SPIFFE-compatible and use SPIFFE-IDs rather than tied to DNS names.

> What are the general concerns here? Is it that widening what's allowed on this profile might make it harder to federate trust domains?

Yeah, you're asking to add a new configuration option that _may or may not be_ required, depending on who you are trying to federate with. I'm not 100% opposed, but we need to weigh any use cases we unlock against the extra ambiguity and complexity.

Is there some concrete system or product you are trying to build? If so, can you describe it?

**Comment by evan2645** [comment]
We discussed this issue in SIG-Spec call last week , so wanted to post a summary here

> Yeah, you're asking to add a new configuration option that may or may not be required, depending on who you are trying to federate with.

This was also raised by call participants. The intention of the `https_web` profile is to "just work" based on public trust, which is why the profile spec is shaped the way it is.

The web pki ecosystem is a specific one. It widely assumes a broad bucket of implicit public trust. SPIFFE is different in that trust is explicit. So to me, it feels like this request is less a problem with the profile and more a problem with the web pki patterns in general. But putting that aside for a moment, consensus on the call was that we should not introduce new optional configurables to the `https_web` profile. Instead, the favored path to introduce this would be via a new profile, which probably leads to a conversation on whether or not this use case justifies a new/dedicated profile (or if there are other use cases as well which would additionally justify the existence of an add'l profile)

We didn't get to that last part on the call 😂 but we did get far enough to get consensus on the direction being a new profile, if warranted.

**Comment by kfox1111** [comment]
The OS provided web chain of trust though is often directly modified at various sites to inject their local CA's. So, it kind if is possible today. On top of that, adding the option to restrict the endpoint to just one CA, is a security optimization that only increases security?

**Comment by sorindumitru** [comment]
> We're talking about SPIFFE Federation here, so the assumption is that you already have some SPIFFE trust domain with a CA root of trust. That CA signs an SVID that is issued to the bundle endpoint. Then to authenticate the endpoint, you need the URL, SPIFFE-ID and CA --- very similar to what you're asking for except the certs are SPIFFE-compatible and use SPIFFE-IDs rather than tied to DNS names.

I think what you're describing here is the `https_spiffe` profile. I can't use that because not all trust domains can directly talk to each other. Because of that, I chose to go through a hub to fetch federated trust bundles. While this could have a SPIFFE SVID, I feel like it's not the best option. It makes a trust domain more "powerful/special" by being able to issue SVIDs for that trust domain and it would also require some bootstrapping for the initial trust of that trust domain.

> The OS provided web chain of trust though is often directly modified at various sites to inject their local CA's. So, it kind if is possible today. On top of that, adding the option to restrict the endpoint to just one CA, is a security optimization that only increases security?

Yes, that's true and it's in fact what I'm doing today. But I'd like to go even further here and have the CA only issue certificates for the DNS SANs of the bundle endpoints I have. Naturally, this CA shouldn't be installed in system trust stores since then you'd be trusting it to issue certificates for any DNS SAN.

I think having a separate profile for this is going a bit overboard since there's not enough difference between the two. And as Keving mentioned, you kind of are able to switch between the two depending on what's in the system trust store. I think explicitly warning about the difficulties you may encounter when setting this up might be enough.

**Comment by spikecurtis** [comment]
You could have the "hub" be a small, separate trust domain with literally only one SPIFFE-ID: that of the bundle endpoint server. The bundle endpoint server doesn't need to share a trust domain with any of the bundles it serves.

You still have your CA and a leaf certificate for the bundle endpoint sever --- just SPIFFE-compatible for both. You don't need to install the CA in any system trust stores and can use the CA _just_ for authenticating against the bundle endpoint.

**Comment by sorindumitru** [comment]
Yes, I thought that's what you might be proposing. I think that's a bit overkill. Instead of having to managed a CA, now I have to manage that + a trust domain + making sure any other trust domain bootstraps its federation with it and successfully maintains it when rotating the CA. It would be much simpler if I only had that CA.

Do we not think that warning users about the potential pitfalls of this would be enough to address our concerns? As I mentioned in my comment, there's nothing preventing users from doing this today, they just have to go via the system trust store. I'd rather have a better option for them and warn them about any future issues. It's also something that they can backtrack on if they actually need the functionality.

I'm also somewhat doubtful of how big of an issue this is/will be. I can see it becoming an issue if you have federation between completely different entities, e.g. company A and company B, but my gut feeling is that there isn't that much of this going on.

**Comment by sorindumitru** [comment]
Another point is that even with plain PKI, the set of trusted CAs varies between systems, so you may end up not being able to federate because of this. This can happen even with large certificate authorities since they ocassionally have to create new CAs which may not yet trusted by all systems. Machines should have an up to date list of CAs, but it's not always the case. That's not an issue I want to be solving when setting up federation.
